src/gha_runner/__main__.py

```diff
@@ -1,13 +1,107 @@
+def env_parse_helper(
+    params: dict, var: str, key: str, is_json: bool = False
+) -> dict:
+    val = os.environ.get(var)
+    if val is not None:
+        if val != "":
+            if is_json:
+                params[key] = json.loads(val)
+            else:
+                params[key] = val
+    return params
+
+
+def parse_aws_params() -> dict:
+    params = {}
+    ami = os.environ.get("INPUT_AWS_IMAGE_ID")
+    if ami is not None:
+        params["image_id"] = ami
+    instance_type = os.environ.get("INPUT_AWS_INSTANCE_TYPE")
+    if instance_type is not None:
+        params["instance_type"] = instance_type
+    params = env_parse_helper(params, "INPUT_AWS_SUBNET_ID", "subnet_id")
+    params = env_parse_helper(
+        params, "INPUT_AWS_SECURITY_GROUP_ID", "security_group_id"
+    )
+    params = env_parse_helper(params, "INPUT_AWS_IAM_ROLE", "iam_role")
+    params = env_parse_helper(params, "INPUT_AWS_TAGS", "tags", is_json=True)
+    region_name = os.environ.get("INPUT_AWS_REGION_NAME")
+    if region_name is not None:
+        params["region_name"] = region_name
+    home_dir = os.environ.get("INPUT_AWS_HOME_DIR")
+    if home_dir is not None:
+        params["home_dir"] = home_dir
+    params = env_parse_helper(params, "INPUT_AWS_LABELS", "labels")
+    return params
+
+
+def get_instance_mapping() -> dict[str, str]:
+    mapping_str = os.environ.get("INPUT_INSTANCE_MAPPING")
+    if mapping_str is None:
+        raise ValueError(
+            "Missing required input variable INPUT_INSTANCE_MAPPING"
+        )
+    return json.loads(mapping_str)
+
+
+def start_runner_instances(
+    provider: str, cloud_params: dict, gh: GitHubInstance, count: int
+):
+    release = gh.get_latest_runner_release(platform="linux", architecture="x64")
+    cloud_params["runner_release"] = release
+    print("Starting up...")
+    # Create a GitHub instance
+    print("Creating GitHub Actions Runner")
+    # We need to create a runner token first
+    runner_tokens = gh.create_runner_tokens(count)
+    cloud_params["gh_runner_tokens"] = runner_tokens
+    # We will also get the latest runner release
+    cloud = CloudDeploymentFactory().get_provider(
+        provider_name=provider, **cloud_params
+    )
+    mappings = cloud.create_instances()
+    instance_ids = list(mappings.keys())
+    github_labels = list(mappings.values())
+    with open(os.environ["GITHUB_OUTPUT"], "a") as output:
+        json_mappings = json.dumps(mappings)
+        output.write(f"mapping={json_mappings}\n")
+    print("Waiting for instance to be ready...")
+    cloud.wait_until_ready(instance_ids)
+    print("Instance is ready!")
+    # gh.wait_for_runner("testing")
+    for label in github_labels:
+        gh.wait_for_runner(label)
+
+
+def stop_runner_instances(
+    provider: str, cloud_params: dict, gh: GitHubInstance
+):
+    print("Shutting down...")
+    try:
+        mappings = get_instance_mapping()
+    except Exception as e:
+        print(e)
+        return
```

Comment:
Under what circumstance would we hit an exception in this? I'm trying to figure out whether failing somewhat silently (as here, with a non-error return) is the right thing or whether this should error and fail out.

I know that we usually don't want our code to cause a user's CI to go red, but if there's a risk of leaving an EC2 instance racking up bills, we probably should.

Reply:
This fails when we do not provide any instance mappings to stop the runner (basically just that the environment variable is not set). In this case, if it isn't set, we can't do anything anyways. There is a risk that this does leave an instance open because of misconfiguration.

Reply:
With that in mind, we probably should cause an error here and return a different return code.

Reply:
Fixed in [4116b1e](https://github.com/omsf-eco-infra/gha-runner/pull/8/commits/4116b1ee7455514ff9743657cec28f9bd984d1cf)